### PR TITLE
[SHIPA-1866] Introduce nginx ingress controller

### PR DESCRIPTION
**Pull request by koncar**
# Description

Add support for Nginx ingress controller

To test, install nginx 
```
helm repo add ingress-nginx https://kubernetes.github.io/ingress-nginx
helm repo update

helm install ingress-nginx ingress-nginx/ingress-nginx
```

Add nginx framework
```
ketch framework add fw2-nginx --ingress-type nginx --ingress-service-endpoint {ip of nginx service}
```
Deploy app
```
ketch app deploy go1 -i shipasoftware/go-app:v1 -k fw1-nginx
```

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] Chore (documentation addition or typo, file relocation)

## Testing

- [X] New tests were added with this PR that prove my fix is effective or that my feature works (describe below this bullet)
- [ ] This change requires no testing (i.e. documentation update)

## Documentation

- [X] All added public packages, funcs, and types have been documented with doc comments
- [X] I have commented my code, particularly in hard-to-understand areas

## Final Checklist:

- [X] I followed standard [GitHub flow](https://guides.github.com/introduction/flow/) guidelines
- [X] I have performed a self-review of my own code
- [X] My changes generate no new warnings

### CODE REVIEW

File: internal/chart/testdata/charts/dashboard-nginx.yaml [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+            - name: VAR
+              value: VALUE
+          image: shipasoftware/go-app:v2
+          ports:
+          - containerPort: 9091
+      imagePullSecrets:
+            - name: default-image-pull-secret
+---
+# Source: dashboard/templates/ingress.yaml
+apiVersion: networking.k8s.io/v1beta1
```
Comment (aleksej-paschenko) [inline internal/chart/testdata/charts/dashboard-nginx.yaml:319]: it should be v1 instead v1beta1
https://kubernetes.io/docs/reference/using-api/deprecation-guide/#ingress-v122

File: internal/templates/nginx/yamls/ingress.yaml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+---
+{{- end }}
+{{- end }}
+{{- end }}
+
+{{- if .Values.app.isAccessible }}
+{{- if .Values.app.ingress.https }}
+{{- range $i, $deployment := .Values.app.deployments }}
+{{- if gt $deployment.routingSettings.weight 0.0}}
+apiVersion: networking.k8s.io/v1beta1
```
Comment (aleksej-paschenko) [inline internal/templates/nginx/yamls/ingress.yaml:46]: `networking.k8s.io/v1`
https://kubernetes.io/docs/reference/using-api/deprecation-guide/#ingress-v122

File: internal/templates/nginx/yamls/ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,81 @@
+{{- if .Values.app.isAccessible }}
+{{- if .Values.app.ingress.http }}
+{{- range $i, $deployment := .Values.app.deployments }}
+{{- if gt $deployment.routingSettings.weight 0.0}}
+apiVersion: networking.k8s.io/v1beta1
```
Comment (aleksej-paschenko) [inline internal/templates/nginx/yamls/ingress.yaml:5]: `networking.k8s.io/v1`
https://kubernetes.io/docs/reference/using-api/deprecation-guide/#ingress-v122

**Review by aleksej-paschenko — CHANGES_REQUESTED** [review]

**Review by DavisFrench — APPROVED** [review]
I tested out the nginx ingress controller and everything seemed to work fine. All the tests pass and the code looks good.

**Comment by stinkyfingers** [comment]
I agree w/ Aleksej and it looks like you need another service in your test chart, but looks good otherwise: 
```
---
# Source: dashboard/templates/gateway_service.yaml
apiVersion: v1
kind: Service
metadata:
  labels:
    theketch.io/app-name: dashboard
    theketch.io/is-isolated-run: "false"
  name: app-dashboard
spec:
  type: ClusterIP
  ports:
    - name: http-default-1
      port: 9091
      protocol: TCP
      targetPort: 9091
  selector:
    theketch.io/app-name: dashboard
    theketch.io/app-process: web
    theketch.io/app-deployment-version: "4"
    theketch.io/is-isolated-run: "false"
---
```

File: internal/templates/nginx/yamls/ingress.yaml (hunk 3 above)
Comment (koncar) [reply internal/templates/nginx/yamls/ingress.yaml]: agree modified

File: internal/templates/nginx/yamls/ingress.yaml (hunk 2 above)
Comment (koncar) [reply internal/templates/nginx/yamls/ingress.yaml]: agree, modified

**Comment by koncar** [comment]
> I agree w/ Aleksej and it looks like you need another service in your test chart, but looks good otherwise:
> 
> ```
> ---
> # Source: dashboard/templates/gateway_service.yaml
> apiVersion: v1
> kind: Service
> metadata:
>   labels:
>     theketch.io/app-name: dashboard
>     theketch.io/is-isolated-run: "false"
>   name: app-dashboard
> spec:
>   type: ClusterIP
>   ports:
>     - name: http-default-1
>       port: 9091
>       protocol: TCP
>       targetPort: 9091
>   selector:
>     theketch.io/app-name: dashboard
>     theketch.io/app-process: web
>     theketch.io/app-deployment-version: "4"
>     theketch.io/is-isolated-run: "false"
> ---
> ```

yup fixed after rebase

File: internal/templates/nginx/yamls/ingress.yaml [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+{{- if .Values.app.ingress.http }}
+{{- range $i, $deployment := .Values.app.deployments }}
+{{- if gt $deployment.routingSettings.weight 0.0}}
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: {{ $.Values.app.name }}-{{ $i }}-http-ingress
+  annotations:
+    {{- if $.Values.ingressController.className }}
+    kubernetes.io/ingress.class: {{ $.Values.ingressController.className }}
```
Comment (aleksej-paschenko) [inline internal/templates/nginx/yamls/ingress.yaml:11]: let's use `quote` function there
```
    kubernetes.io/ingress.class: {{ $.Values.ingressController.className | quote }}
```
we will fail to render a chart if className is  `true` or `1`. 
I'll be a valid k8s manifest but helm doesn't like such annotations/labels

**Review by aleksej-paschenko — CHANGES_REQUESTED** [review]

**Review by aleksej-paschenko — APPROVED** [review]

File: internal/templates/nginx/yamls/ingress.yaml (hunk 4 above)
Comment (koncar) [reply internal/templates/nginx/yamls/ingress.yaml:11]: lets fix in another PR

**Pull request merged**